<a href="https://colab.research.google.com/github/sv2639/proj3/blob/main/proj3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem 1: Explore Different Modes of Operation Through Manual Encryption and Decryption

## Problem 1: Encryption Modes

### Part 2a - Mode: ECB (Electronic Codebook)
Convert the plaintext **FOO** into binary using the provided table:
- **F** = 0101  
- **O** = 1110  
- **O** = 1110  

Apply the block cipher **Ek** to each block:
- **Ek(0101) = 1001**  
- **Ek(1110) = 1110**  
- **Ek(1110) = 1110**  

Convert the ciphertext binary back into letters using the table:
- **1001 = J**  
- **1110 = O**  

Final output:  
**FOO → JOO**  

---

### Part 2b - Mode: CBC (Cipher Block Chaining) with IV = 1010
#### Step 1: Encrypt each block  
1. **Block 1**:  
   - XOR with Initialization Vector:  
     - **0101 ⊕ 1010 = 1111**  
   - Apply the cipher **Ek** on the block:  
     - **Ek(1111) = 1111**  

2. **Block 2**:  
   - XOR with previous ciphertext:  
     - **1110 ⊕ 1111 = 0001**  
   - Apply **Ek**:  
     - **Ek(0001) = 0001**  

3. **Block 3**:  
   - XOR with previous ciphertext:  
     - **1110 ⊕ 0001 = 1111**  
   - Apply **Ek**:  
     - **Ek(1111) = 1111**  

Convert **1111 0001 1111** back into letters using the table:  
**FOO → PBP**  

---

### Part 2c - Mode: CTR (Counter) with ctr = 1010
1. Encrypt the counter:  
   - **CTR = 1010**  
   - Apply **Ek**:  
     - **Ek(1010) = 0110**  

2. Encrypt each block:  
   - **Block 1**:  
     - **0101 ⊕ 0110 = 0011**  

   - **Increase counter**:  
     - **CTR = 1011**  
     - **Ek(1011) = 0111**  

   - **Block 2**:  
     - **1110 ⊕ 0111 = 1001**  

   - **Increase counter**:  
     - **CTR = 1100**  
     - **Ek(1100) = 1010**  

   - **Block 3**:  
     - **1110 ⊕ 1010 = 0100**  

Convert **0011 1001 0100** back into letters using the table:  
**FOO → DJE**  

---

## Problem 1: Decryption Task

### 1) ECB (Electronic Codebook)
Convert the ciphertext into binary:
- **J** = 1001  
- **O** = 1110  
- **O** = 1110  

Apply the inverse of the encryption function:
- **1001 → 0101 → F**  
- **1110 → 1110 → O**  
- **1110 → 1110 → O**  

Final decrypted output:  
**JOO → FOO**  

---

### 2) CBC (Cipher Block Chaining) with IV = 1010
Convert the ciphertext to binary:
- **P** = 1111  
- **B** = 0001  
- **P** = 1111  

Decrypt using CBC:
1. XOR with Initialization Vector:  
   - **1010 ⊕ 1111 = 0101**  
   - Apply **Ek⁻¹**:  
     - **1001 → F**  

2. **Block 2**:  
   - XOR with previous ciphertext:  
     - **0001 ⊕ 1111 = 1110**  
   - Apply **Ek⁻¹**:  
     - **1110 → O**  

3. **Block 3**:  
   - XOR with previous ciphertext:  
     - **0001 ⊕ 1111 = 1110**  
   - Apply **Ek⁻¹**:  
     - **1110 → O**  

Final decrypted output:  
**PBP → FOO**  

---

### 3) CTR (Counter Mode) with Counter = 1010
Convert ciphertext into binary:
- **DJE** = **0011 1001 0100**  

Counters:  
- **Counter1** = **1010**  
- **Counter2** = **1011**  
- **Counter3** = **1100**  

Decrypt each block:
1. **Encrypt Counter1**:  
   - **Ek(1010) = 0110**  
   - **c1 = D = 0011**  
   - **o1 ⊕ c1 = 0110 ⊕ 0011 = 0101 (F)**  

2. **Encrypt Counter2**:  
   - **Ek(1011) = 0111**  
   - **c2 = J = 1001**  
   - **o2 ⊕ c2 = 0111 ⊕ 1001 = 1110 (O)**  

3. **Encrypt Counter3**:  
   - **Ek(1100) = 1010**  
   - **c3 = E = 0100**  
   - **o3 ⊕ c3 = 1010 ⊕ 0100 = 1110 (O)**  

Final decrypted output:  
**DJE → FOO**  


# Problem 2

 In the pseudo code for question 2 in project 3, it says the following:
> For the project, we need to modify the pseudo code to fit our class project. The pseudo-code for this
> mini block cipher based on the SAES is as follows:
> 1. Get (Key0, Key1, Key2) from Key K using the key expansion


It says to get key0, key1, and key2

But then steps 2 and 3 say to use key1 and key2, completely skipping key0

The pseudo-code for SAES says to use key0, before either of the encrypt rounds. Meaning key0 actually gets used.

But I don't see where key0 gets used in the mini block cipherSo that's why I'm messaging you.Why does key0 get made if it's not being used?

Secondly. For the substitute portion of the algorithm are we using the substitution box from the slides of this week's lecture?



## Task 1A

In [30]:
# KEY EXPANSION


import sys

# AES S-Box (from standard AES specification)
s_box = [
    0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5, 0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
    0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0, 0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
    0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC, 0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
    0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A, 0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
    0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0, 0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
    0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B, 0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
    0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85, 0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
    0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5, 0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
    0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17, 0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
    0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88, 0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
    0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C, 0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
    0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9, 0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
    0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6, 0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
    0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E, 0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
    0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94, 0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
    0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68, 0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16
]

# Generate Inverse S-Box
inv_s_box = [0] * 256
for i in range(256):
    inv_s_box[s_box[i]] = i


Rcon = [0x21, 0x13, 0x37, 0x7F, 0xAC, 0x99]


def sub_byte(byte):
    return s_box[byte]

def inv_sub_byte(byte):
    return inv_s_box[byte]

def rot_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF  # Least significant byte
    byte2 = (word_16bit >> 8) & 0xFF # Most significant byte
    return (byte2 << 8) | byte1 # Swap bytes

def rot_word_8bit(word_8bit):
    byte1 = word_8bit & 0xF  # Least significant byte
    byte2 = (word_8bit >> 4) & 0xF # Most significant byte
    return (byte2 << 4) | byte1 # Swap bytes

def sub_word_8bit(word_8bit):
    byte1 = word_8bit & 0xF
    byte2 = (word_8bit >> 4) & 0xF
    # Apply the substitution box and make sure the result is within 8 bits
    substituted = (sub_byte(byte2) << 4) | sub_byte(byte1)

    return substituted & 0xFF  # Mask the result to 8 bits

def sub_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (sub_byte(byte2) << 8) | sub_byte(byte1)


def inv_sub_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (inv_sub_byte(byte2) << 8) | inv_sub_byte(byte1)

def key_expansion(key):
    """
    Simplified key expansion for a 16-bit key to generate 3 round keys (each 16-bit): Key0, Key1, Key2.
    key_16bit: 16-bit integer representing the AES key.
    Returns: A list of 3 round keys [Key0, Key1, Key2] (each 16-bit integer).
    """
    words = [0] * 6 # We need 6 words W0, W1, W2, W3, W4, W5, which will be Key0, Key1, Key2
    words[0] = (key >> 8) & 0xFF  # Initialize first word with the first 8-bits of the key
    words[1] = key & 0xFF   # Initialize second word with the second 8-bits of the key

    for i in range(2, 6): # Generate words W2 to W5
        temp = words[i-1]
        temp_rot = rot_word_8bit(temp)
        temp_sub = sub_word_8bit(temp_rot)
        rc_val = Rcon[i-2]
        #rc_word = (rc_val << 4) & 0xFF  # Ensure it's still an 8-bit value
        words[i] = words[i-1] ^ temp_sub ^ rc_val
    """
    round_keys = []
    for i in range(6): # Generate 3 round keys Key0 to Key2
        round_keys.append(words[i]) # Keyi = Wi
    """
    # Combine two 8-bit words into 16-bit words for key0, key1, key2
    key0 = (words[0] << 8) | words[1]  # Combine w0 and w1 into key0 (16 bits)
    key1 = (words[2] << 8) | words[3]  # Combine w2 and w3 into key1 (16 bits)
    key2 = (words[4] << 8) | words[5]  # Combine w4 and w5 into key2 (16 bits)

    return key0, key1, key2
"""
def key_expansion(key_16bit):

    Simplified key expansion for a 16-bit key to generate 3 round keys (each 16-bit): Key0, Key1, Key2.
    key_16bit: 16-bit integer representing the AES key.
    Returns: A list of 3 round keys [Key0, Key1, Key2] (each 16-bit integer).

    words = [0] * 3 # We need 3 words W0, W1, W2 which will be Key0, Key1, Key2
    words[0] = key_16bit  # Initialize first word with the key

    for i in range(1, 3): # Generate words W1 and W2
        temp = words[i-1]
        temp_rot = rot_word_16bit(temp)
        temp_sub = sub_word_16bit(temp_rot)
        rc_val = Rcon[i-1] if i <= len(Rcon) else 0 # Get Round Constant for round i (or 0 if out of Rcon range)
        rc_word = rc_val << 8 # Represent RC as 16-bit word (pad with zero byte in lower byte)
        words[i] = words[i-1] ^ temp_sub ^ rc_word

    round_keys = []
    for i in range(3): # Generate 3 round keys Key0 to Key2
        round_keys.append(words[i]) # Keyi = Wi
    return round_keys

"""
example_key_16bit = 0xABCD # Example 16-bit key

round_keys = key_expansion(example_key_16bit)

print("Round Keys (16-bit hex) - Key0, Key1, Key2:")
print(f"Key0 = 0x{round_keys[0]:04X}")
print(f"Key1 = 0x{round_keys[1]:04X}")
print(f"Key2 = 0x{round_keys[2]:04X}")


Round Keys (16-bit hex) - Key0, Key1, Key2:
Key0 = 0xABCD
Key1 = 0x1BE3
Key2 = 0x2F26


In [4]:
#ENCRYPTION


def subsitute_state(state_16bit):
    return sub_word_16bit(state_16bit)

def shift_state(state_16bit):
    return rot_word_16bit(state_16bit)

def mix_state(state_16bit):
    byte1 = state_16bit & 0xFF
    byte2 = (state_16bit >> 8) & 0xFF
    mixed_byte1 = byte2
    mixed_byte2 = byte1 ^ byte2
    return (mixed_byte2 << 8) | mixed_byte1

def add_round_key(state_16bit, round_key_16bit):
    return state_16bit ^ round_key_16bit

def encrypt_round1(plaintext_16bit, key1_16bit):
    state = plaintext_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = mix_state(state)
    state = add_round_key(state, key1_16bit)
    return state # Intermediate state X

def encrypt_round2(intermediate_state_16bit, key2_16bit):
    state = intermediate_state_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = add_round_key(state, key2_16bit) # No Mix in the last round
    return state # Ciphertext


# Combined Encryption function
def encrypt(plaintext_16bit, key_16bit):
    """
    Encrypts a 16-bit plaintext using the simplified 2-round SAES with a 16-bit key.
    plaintext_16bit: 16-bit integer plaintext.
    key_16bit: 16-bit integer key.
    Returns: 16-bit integer ciphertext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 for encryption rounds

    intermediate_state_X = encrypt_round1(plaintext_16bit, key1)
    ciphertext_16bit = encrypt_round2(intermediate_state_X, key2)
    return ciphertext_16bit

In [5]:
#DECRYPTION


def inverse_substitute_state(state_16bit):
    return inv_sub_word_16bit(state_16bit)

def inverse_shift_state(state_16bit):
    return shift_state(state_16bit) # ShiftRows is its own inverse for byte swap

def inverse_mix_state(state_16bit):
    # If mix_state is simplified to: mixed_byte1 = byte1 ^ byte2; mixed_byte2 = byte2;
    # then inverting it: byte2 = mixed_byte2; byte1 = mixed_byte1 ^ byte2 = mixed_byte1 ^ mixed_byte2;
    mixed_byte1 = state_16bit & 0xFF
    mixed_byte2 = (state_16bit >> 8) & 0xFF
    byte2 = mixed_byte1
    byte1 = mixed_byte2 ^ byte2
    return (byte2 << 8) | byte1


def decrypt_round2(ciphertext_16bit, key2_16bit): # Decryption round 2 (first decryption round)
    state = ciphertext_16bit
    state = add_round_key(state, key2_16bit)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Intermediate state Y

def decrypt_round1(intermediate_state_16bit, key1_16bit): # Decryption round 1 (second decryption round)
    state = intermediate_state_16bit
    state = add_round_key(state, key1_16bit)
    state = inverse_mix_state(state)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Plaintext

# Combined Decryption function
def decrypt(ciphertext_16bit, key_16bit):
    """
    Decrypts a 16-bit ciphertext using the simplified 2-round SAES with a 16-bit key.
    ciphertext_16bit: 16-bit integer ciphertext.
    key_16bit: 16-bit integer key (same key used for encryption).
    Returns: 16-bit integer recovered plaintext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 (same keys as in encryption, but in reverse round order in decryption process)

    intermediate_state_Y = decrypt_round2(ciphertext_16bit, key2) # Decryption round 2 uses Key2
    recovered_plaintext_16bit = decrypt_round1(intermediate_state_Y, key1) # Decryption round 1 uses Key1
    return recovered_plaintext_16bit

In [9]:
import random

# Function to generate a random 16-bit key
def generate_random_key_16bit():
    """
    Generates a random 16-bit key.
    Returns: A random 16-bit integer.
    """
    return random.randint(0, 0xFFFF)

In [10]:

# Example usage with a 16-bit key and plaintext
example_key_16bit = generate_random_key_16bit()
plaintext_16bit = 0x1234 # Example 16-bit plaintext

print("Plaintext:", f"0x{plaintext_16bit:04X}")
print("Key:", f"0x{example_key_16bit:04X}")

ciphertext = encrypt(plaintext_16bit, example_key_16bit)
print("\nCiphertext:", f"0x{ciphertext:04X}")

recovered_plaintext = decrypt(ciphertext, example_key_16bit)
print("Recovered Plaintext:", f"0x{recovered_plaintext:04X}")

print("\nVerification:")
if plaintext_16bit == recovered_plaintext:
  print("\n Encryption and Decryption successful! Recovered plaintext matches original plaintext.")
else:
  print("\n Encryption and Decryption failed. Recovered plaintext does NOT match original plaintext.")

Plaintext: 0x1234
Key: 0xE2F6

Ciphertext: 0xEB74
Recovered Plaintext: 0x1234

Verification:

 Encryption and Decryption successful! Recovered plaintext matches original plaintext.


The above code is likely the final version of Task 1a. Unless someone thinks that we should add lower case letters/numbers/special chars. Which could be acheived with slight modification. But I'm not sure it's needed.

Why I think it's the final version:


*   16 bit key
*   Full alphabet
*   Spaces allowed
*   Works without issue





## Task 1B

In [11]:
# Generate a random 16-bit key
example_key_16bit = generate_random_key_16bit()

print("Using Random Key:", f"0x{example_key_16bit:04X}")

plaintext_ciphertext_pairs = {} # Dictionary to store pairs

print("\nGenerating 10 Plaintext-Ciphertext pairs with random key:")
print("---------------------------------------")

for i in range(10):
    # Generate sequential plaintexts for simplicity, you can use random plaintexts as well
    plaintext_16bit = generate_random_key_16bit()
    plaintext_hex = f"0x{plaintext_16bit:04X}"

    ciphertext = encrypt(plaintext_16bit, example_key_16bit)
    ciphertext_hex = f"0x{ciphertext:04X}"

    print(f"Pair {i+1}:")
    print(f"  Plaintext:  {plaintext_hex}")
    print(f"  Ciphertext: {ciphertext_hex}")
    print("---------------------------------------")

    # Store the pair in the dictionary
    plaintext_ciphertext_pairs[f"Pair {i+1}"] = {
        "plaintext": plaintext_hex,
        "ciphertext": ciphertext_hex
    }

print("\nVerification of Decryption for the last pair:")
last_plaintext = plaintext_16bit
last_ciphertext = ciphertext
recovered_plaintext = decrypt(last_ciphertext, example_key_16bit)

print(f"Ciphertext to decrypt: 0x{last_ciphertext:04X}")
print(f"Recovered Plaintext:   0x{recovered_plaintext:04X}")

if last_plaintext == recovered_plaintext:
    print("\n Decryption Verified for the last pair: Recovered plaintext matches original.")
else:
    print("\n Decryption Verification Failed for the last pair.")

print("\nPlaintext-Ciphertext Pairs Dictionary:")
print(plaintext_ciphertext_pairs)

Using Random Key: 0xF2ED

Generating 10 Plaintext-Ciphertext pairs with random key:
---------------------------------------
Pair 1:
  Plaintext:  0x8956
  Ciphertext: 0xC617
---------------------------------------
Pair 2:
  Plaintext:  0x6072
  Ciphertext: 0xE84B
---------------------------------------
Pair 3:
  Plaintext:  0x21EB
  Ciphertext: 0xA25C
---------------------------------------
Pair 4:
  Plaintext:  0xB884
  Ciphertext: 0x862E
---------------------------------------
Pair 5:
  Plaintext:  0x89C3
  Ciphertext: 0xCB17
---------------------------------------
Pair 6:
  Plaintext:  0xC5B6
  Ciphertext: 0xFE06
---------------------------------------
Pair 7:
  Plaintext:  0x258C
  Ciphertext: 0x06D1
---------------------------------------
Pair 8:
  Plaintext:  0x630A
  Ciphertext: 0x2C6A
---------------------------------------
Pair 9:
  Plaintext:  0x1413
  Ciphertext: 0x1F33
---------------------------------------
Pair 10:
  Plaintext:  0x48FF
  Ciphertext: 0xE599
---------------

## Task 2A

In [12]:
def mitm_attack(plaintext_16bit1, ciphertext_16bit1, plaintext_16bit2, ciphertext_16bit2):
    """
    Performs a Meet-in-the-Middle attack to find the key pair (Key1, Key2)
    for the simplified 2-round SAES, using TWO plaintext-ciphertext pairs for verification.
    plaintext_16bit1: First known plaintext (16-bit integer).
    ciphertext_16bit1: Corresponding ciphertext for first plaintext (16-bit integer).
    plaintext_16bit2: Second known plaintext (16-bit integer).
    ciphertext_16bit2: Corresponding ciphertext for second plaintext (16-bit integer).
    Returns: A list of potential key pairs [(Key1, Key2), ...] that could have encrypted
             BOTH plaintext pairs to their respective ciphertexts.
    """
    x_values = {} # Dictionary to store X = encrypt_round1(Key1, P1) values (using the first plaintext)
    x_prime_values = {} # Dictionary to store X' = decrypt_round2(Key2, C1) values (using the first ciphertext)

    print("Step A: Pre-calculating X values (using P1)...")
    for key1 in range(0x10000): # Iterate through all possible 16-bit Key1 values
        x = encrypt_round1(plaintext_16bit1, key1)
        if x not in x_values:
            x_values[x] = []
        x_values[x].append(key1)

    print("Step B: Pre-calculating X' values (using C1)...")
    for key2 in range(0x10000): # Iterate through all possible 16-bit Key2 values
        x_prime = decrypt_round2(ciphertext_16bit1, key2)
        if x_prime not in x_prime_values:
            x_prime_values[x_prime] = []
        x_prime_values[x_prime].append(key2)

    matched_key_pairs = []
    print("Step C: Finding matching key pairs...")
    for x_val, key1_list in x_values.items():
        if x_val in x_prime_values:
            for key1 in key1_list:
                for key2 in x_prime_values[x_val]:
                    matched_key_pairs.append((key1, key2))

    print(f"Step C: Found {len(matched_key_pairs)} potential key pairs.")
    return matched_key_pairs



def verify_key_pair(key_pair, plaintext_16bit1, ciphertext_16bit1, plaintext_16bit2, ciphertext_16bit2):
    """
    Verifies if a key pair correctly encrypts BOTH plaintext pairs to their respective ciphertexts.
    key_pair: A tuple (Key1, Key2).
    plaintext_16bit1: First plaintext (16-bit integer).
    ciphertext_16bit1: First ciphertext (16-bit integer).
    plaintext_16bit2: Second plaintext (16-bit integer).
    ciphertext_16bit2: Second ciphertext (16-bit integer).
    Returns: True if the key pair is correct for both pairs, False otherwise.
    """
    key1, key2 = key_pair
    calculated_intermediate_X1 = encrypt_round1(plaintext_16bit1, key1)
    calculated_ciphertext1 = encrypt_round2(calculated_intermediate_X1, key2)
    calculated_intermediate_X2 = encrypt_round1(plaintext_16bit2, key1)
    calculated_ciphertext2 = encrypt_round2(calculated_intermediate_X2, key2)
    return calculated_ciphertext1 == ciphertext_16bit1 and calculated_ciphertext2 == ciphertext_16bit2


## Task 2B

In [27]:
# Example Usage of MITM Attack with TWO P-C pairs

# 1. Generate a random key
true_key_16bit = generate_random_key_16bit()

# 2. Generate TWO plaintext-ciphertext pairs using the same key
plaintext_example1 = 0xABCD
ciphertext_example1 = encrypt(plaintext_example1, true_key_16bit)

plaintext_example2 = 0x5678 # A different plaintext
ciphertext_example2 = encrypt(plaintext_example2, true_key_16bit)


print("True Key:", f"0x{true_key_16bit:04X}")
print("\nPlaintext 1:", f"0x{plaintext_example1:04X}")
print("Ciphertext 1:", f"0x{ciphertext_example1:04X}")
print("\nPlaintext 2:", f"0x{plaintext_example2:04X}")
print("Ciphertext 2:", f"0x{ciphertext_example2:04X}")

# 3. Perform MITM attack using the TWO plaintext-ciphertext pairs
print("\nStarting Meet-in-the-Middle Attack (using two P-C pairs)...")
potential_key_pairs = mitm_attack(plaintext_example1, ciphertext_example1, plaintext_example2, ciphertext_example2)

# 4. Verify potential key pairs against BOTH P-C pairs
print("\nStep D: Verifying potential key pairs against both P-C pairs...")
if potential_key_pairs:
    print(f"Found {len(potential_key_pairs)} potential key pairs. Verifying...")
    correct_key_pair_found = None
    for key_pair in potential_key_pairs:
        if verify_key_pair(key_pair, plaintext_example1, ciphertext_example1, plaintext_example2, ciphertext_example2): # Verify against TWO P-C pairs now
            correct_key_pair_found = key_pair
            break # Assuming we are looking for just one correct key pair, stop after finding the first one
    if correct_key_pair_found:
        key1_found, key2_found = correct_key_pair_found
        print("\n!!!!!! MITM Attack Successful !!!!!!")
        print("Recovered Key Pair:")
        print(f"  Recovered Key1: 0x{key1_found:04X}")
        print(f"  Recovered Key2: 0x{key2_found:04X}")

        # To reconstruct the original key (simplified key expansion is used):
        round_keys_true = key_expansion(true_key_16bit)
        true_key1 = round_keys_true[1]
        true_key2 = round_keys_true[2]
        print("\nVerification with True Keys:")
        print(f"  True Key1:      0x{true_key1:04X}")
        print(f"  True Key2:      0x{true_key2:04X}")

        if key1_found == true_key1 and key2_found == true_key2:
            print("\n!!!!!! Recovered Key Pair matches the true key pair !!!!!!")
        else:
            print("\n?????????? Recovered Key Pair does NOT match the true key pair (but is still a valid key for the given P-C pairs) ??????????")
    else:
        print("\n?????????? MITM Attack Failed: No key pair verified against both P-C pairs (unexpected) ??????????")
else:
    print("\n MITM Attack Failed: No potential key pairs found.")

True Key: 0xC023

Plaintext 1: 0xABCD
Ciphertext 1: 0xFF10

Plaintext 2: 0x5678
Ciphertext 2: 0x8708

Starting Meet-in-the-Middle Attack (using two P-C pairs)...
Step A: Pre-calculating X values (using P1)...
Step B: Pre-calculating X' values (using C1)...
Step C: Finding matching key pairs...
Step C: Found 65536 potential key pairs.

Step D: Verifying potential key pairs against both P-C pairs...
Found 65536 potential key pairs. Verifying...

!!!!!! MITM Attack Successful !!!!!!
Recovered Key Pair:
  Recovered Key1: 0x4B57
  Recovered Key2: 0xDD86

Verification with True Keys:
  True Key1:      0x4B57
  True Key2:      0xDD86

!!!!!! Recovered Key Pair matches the true key pair !!!!!!


## Task 3A

What is the key space for the mini block cipher?


*   The key size for the mini block cipher is 16 bits.
*   This means there are 2^16 or 65,536 possible keys.
*   Since we have two keys (Key1 and Key2), the total key space is:
  *   2^16 × 2^16 = 2^32 = 4,294,967,296 key pairs






## Task 3B

If the mini block cipher is executed twice (double mini block cipher), how many operations are needed for a MITM attack?


* Now, we have a 32-bit key, meaning the key space grows to 2^32.
* The MITM attack splits the key search into two parts:
  * One part encrypts all possible 2^16 values for Key1.
  * The second part decrypts all possible 2^16 values for Key2.
* Number of operations:
  * 2^16 + 2^16 = 2^17 = 131,072 operations

## Task 3C

 If we do exhaustive key search for the double mini block cipher, how many operations are needed?

* Exhaustive key search means testing all 2^32 possible keys.
* Number of operations:
  * 2^32 = 4,294,967,296 operations

## Task 3D

Trade-offs between Meet-in-the-Middle Attack and Exhaustive Key Search

The **Meet-in-the-Middle (MITM) attack** and **exhaustive key search** have distinct trade-offs in terms of speed and memory usage.

#### **1. Speed Trade-off**
- **MITM Attack:** Requires **\(2^{17}\)** operations, which is significantly faster than brute-force.
- **Exhaustive Key Search:** Requires **\(2^{32}\)** operations, making it infeasible for larger key sizes.
- **Advantage:** MITM is much **faster** than brute-force, making it more practical.

#### **2. Memory Trade-off**
- **MITM Attack:** Requires storing **\(2^{16}\)** intermediate values in a lookup table.
- **Exhaustive Key Search:** Requires almost no additional memory beyond key testing.
- **Advantage:** Brute-force requires **less memory**, whereas MITM **trades memory for speed**.

#### **3. Applicability Trade-off**
- **MITM Attack:** Only works for ciphers with a **specific structure** (e.g., two independent rounds).
- **Exhaustive Key Search:** Works for **any** cipher.
- **Advantage:** Brute-force is **more general**, but MITM is **more efficient when applicable**.

#### **Conclusion**
The **MITM attack** is significantly **faster** than exhaustive search, but it requires **additional memory** to store intermediate values. It is a **practical attack** when applicable but does not work on ciphers specifically designed to resist it (e.g., AES). On the other hand, **exhaustive search works universally but is computationally infeasible for large key sizes.**

In [7]:
import sys

# AES S-Box (from standard AES specification)
s_box_orig = [
    0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5, 0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
    0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0, 0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
    0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC, 0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
    0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A, 0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
    0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0, 0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
    0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B, 0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
    0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85, 0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
    0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5, 0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
    0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17, 0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
    0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88, 0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
    0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C, 0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
    0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9, 0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
    0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6, 0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
    0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E, 0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
    0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94, 0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
    0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68, 0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16
]

# Generate Inverse S-Box
inv_s_box_orig = [0] * 256
for i in range(256):
    inv_s_box_orig[s_box_orig[i]] = i


# New S-box from the table
s_box_new = {
    0b0000: 0b1001,
    0b0001: 0b0100,
    0b0010: 0b1010,
    0b0011: 0b1011,
    0b0100: 0b1101,
    0b0101: 0b0001,
    0b0110: 0b1000,
    0b0111: 0b0101,
    0b1000: 0b0110,
    0b1001: 0b0010,
    0b1010: 0b0000,
    0b1011: 0b0011,
    0b1100: 0b1100,
    0b1101: 0b1110,
    0b1110: 0b1111,
    0b1111: 0b0111
}

# Generate Inverse S-Box for the new S-box
inv_s_box_new = {}
for key, value in s_box_new.items():
    inv_s_box_new[value] = key

Rcon = [
    0b10000000,  # Round constant 1 (0x80)
    0b00110000   # Round constant 2 (0x30)
]

def sub_nibble(nibble):
    return s_box_new[nibble]

def inv_sub_nibble(nibble):
    return inv_s_box_new[nibble]

def sub_byte_new(byte):
    high_nibble = (byte >> 4) & 0x0F
    low_nibble = byte & 0x0F
    return (sub_nibble(high_nibble) << 4) | sub_nibble(low_nibble)

def inv_sub_byte_new(byte):
    high_nibble = (byte >> 4) & 0x0F
    low_nibble = byte & 0x0F
    return (inv_sub_nibble(high_nibble) << 4) | inv_sub_nibble(low_nibble)

def rot_word_8bit(word_8bit):
    # Assuming rotation means swapping the two 4-bit nibbles within the 8-bit word
    high_nibble = (word_8bit >> 4) & 0x0F
    low_nibble = word_8bit & 0x0F
    return (low_nibble << 4) | high_nibble

def sub_word_8bit(word_8bit):
    return sub_byte_new(word_8bit)

def inv_sub_word_8bit(word_8bit):
    return inv_sub_byte_new(word_8bit)

def key_expansion(key_16bit):
    """
    Key expansion for a 16-bit key to generate 3 round keys (each 16-bit)
    according to the key schedule in the provided image.
    key_16bit: 16-bit integer representing the AES key.
    Returns: A list of 3 round keys [Key0, Key1, Key2] (each 16-bit integer).
    """
    # Extract 8-bit words W0 and W1
    w0 = (key_16bit >> 8) & 0xFF
    w1 = key_16bit & 0xFF

    # Generate W2
    rot_w1 = rot_word_8bit(w1)
    sub_rot_w1 = sub_word_8bit(rot_w1)
    rcon1 = Rcon[0]
    w2 = w0 ^ sub_rot_w1 ^ (rcon1 >> 8) # Apply Rcon to the high byte

    # Generate W3
    w3 = w2 ^ w1

    # Generate W4
    rot_w3 = rot_word_8bit(w3)
    sub_rot_w3 = sub_word_8bit(rot_w3)
    rcon2 = Rcon[1]
    w4 = w2 ^ sub_rot_w3 ^ (rcon2 >> 8) # Apply Rcon to the high byte

    # Generate W5
    w5 = w4 ^ w3

    # Form 16-bit round keys
    key0 = (w0 << 8) | w1  # W0 || W1
    key1 = (w2 << 8) | w3  # W2 || W3
    key2 = (w4 << 8) | w5  # W4 || W5

    return [key0, key1, key2]

# Example usage with a 16-bit key
example_key_16bit_val = 0b1010011100111011 # Example key from the image (0xA73B)

round_keys = key_expansion(example_key_16bit_val)

print("Round Keys (16-bit hex) - Key0, Key1, Key2:")
print(f"Key0 = 0x{round_keys[0]:04X} (Expected: 0xA73B)")
print(f"Key1 = 0x{round_keys[1]:04X} (Expected: 0x1C27)")
print(f"Key2 = 0x{round_keys[2]:04X} (Expected: 0x7651)")

# Update substitute functions to use the new S-box
def sub_word_16bit_new(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (sub_byte_new(byte2) << 8) | sub_byte_new(byte1)

def inv_sub_word_16bit_new(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (inv_sub_byte_new(byte2) << 8) | inv_sub_byte_new(byte1)

# Update encryption and decryption functions to use the new substitute functions
subsitute_state = sub_word_16bit_new
inverse_substitute_state = inv_sub_word_16bit_new

#ENCRYPTION


def rot_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF   # Least significant byte
    byte2 = (word_16bit >> 8) & 0xFF # Most significant byte
    return (byte2 << 8) | byte1 # Swap bytes

def shift_state(state_16bit):
    return rot_word_16bit(state_16bit)

def mix_state(state_16bit):
    byte1 = state_16bit & 0xFF
    byte2 = (state_16bit >> 8) & 0xFF
    mixed_byte1 = byte2
    mixed_byte2 = byte1 ^ byte2
    return (mixed_byte2 << 8) | mixed_byte1

def add_round_key(state_16bit, round_key_16bit):
    return state_16bit ^ round_key_16bit

def encrypt_round1(plaintext_16bit, key1_16bit):
    state = plaintext_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = mix_state(state)
    state = add_round_key(state, key1_16bit)
    return state # Intermediate state X

def encrypt_round2(intermediate_state_16bit, key2_16bit):
    state = intermediate_state_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = add_round_key(state, key2_16bit) # No Mix in the last round
    return state # Ciphertext


# Combined Encryption function
def encrypt(plaintext_16bit, key_16bit):
    """
    Encrypts a 16-bit plaintext using the simplified 2-round SAES with a 16-bit key.
    plaintext_16bit: 16-bit integer plaintext.
    key_16bit: 16-bit integer key.
    Returns: 16-bit integer ciphertext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 for encryption rounds

    intermediate_state_X = encrypt_round1(plaintext_16bit, key1)
    ciphertext_16bit = encrypt_round2(intermediate_state_X, key2)
    return ciphertext_16bit


#DECRYPTION


def inverse_shift_state(state_16bit):
    return shift_state(state_16bit) # ShiftRows is its own inverse for byte swap

def inverse_mix_state(state_16bit):
    # If mix_state is simplified to: mixed_byte1 = byte1 ^ byte2; mixed_byte2 = byte2;
    # then inverting it: byte2 = mixed_byte2; byte1 = mixed_byte1 ^ byte2 = mixed_byte1 ^ mixed_byte2;
    mixed_byte1 = state_16bit & 0xFF
    mixed_byte2 = (state_16bit >> 8) & 0xFF
    byte2 = mixed_byte1
    byte1 = mixed_byte2 ^ byte2
    return (byte2 << 8) | byte1


def decrypt_round2(ciphertext_16bit, key2_16bit): # Decryption round 2 (first decryption round)
    state = ciphertext_16bit
    state = add_round_key(state, key2_16bit)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Intermediate state Y

def decrypt_round1(intermediate_state_16bit, key1_16bit): # Decryption round 1 (second decryption round)
    state = intermediate_state_16bit
    state = add_round_key(state, key1_16bit)
    state = inverse_mix_state(state)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Plaintext

# Combined Decryption function
def decrypt(ciphertext_16bit, key_16bit):
    """
    Decrypts a 16-bit ciphertext using the simplified 2-round SAES with a 16-bit key.
    ciphertext_16bit: 16-bit integer ciphertext.
    key_16bit: 16-bit integer key (same key used for encryption).
    Returns: 16-bit integer recovered plaintext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 (same keys as in encryption, but in reverse round order in decryption process)

    intermediate_state_Y = decrypt_round2(ciphertext_16bit, key2) # Decryption round 2 uses Key2
    recovered_plaintext_16bit = decrypt_round1(intermediate_state_Y, key1) # Decryption round 1 uses Key1
    return recovered_plaintext_16bit


import random
# Function to generate a random 16-bit key
def generate_random_key_16bit():
    """
    Generates a random 16-bit key.
    Returns: A random 16-bit integer.
    """
    return random.randint(0, 0xFFFF)

# Generate a random 16-bit key
example_key_16bit = generate_random_key_16bit()

print("Using Random Key:", f"0x{example_key_16bit:04X}")

plaintext_ciphertext_pairs = {} # Dictionary to store pairs

print("\nGenerating 10 Plaintext-Ciphertext pairs with random key:")
print("---------------------------------------")

for i in range(10):
    # Generate sequential plaintexts for simplicity, you can use random plaintexts as well
    plaintext_16bit = i + 1
    plaintext_hex = f"0x{plaintext_16bit:04X}"

    ciphertext = encrypt(plaintext_16bit, example_key_16bit)
    ciphertext_hex = f"0x{ciphertext:04X}"

    print(f"Pair {i+1}:")
    print(f"  Plaintext:  {plaintext_hex}")
    print(f"  Ciphertext: {ciphertext_hex}")
    print("---------------------------------------")

    # Store the pair in the dictionary
    plaintext_ciphertext_pairs[f"Pair {i+1}"] = {
        "plaintext": plaintext_hex,
        "ciphertext": ciphertext_hex
    }

print("\nVerification of Decryption for the last pair:")
last_plaintext = plaintext_16bit
last_ciphertext = ciphertext
recovered_plaintext = decrypt(last_ciphertext, example_key_16bit)

print(f"Ciphertext to decrypt: 0x{last_ciphertext:04X}")
print(f"Recovered Plaintext:  0x{recovered_plaintext:04X}")

if last_plaintext == recovered_plaintext:
    print("\n Decryption Verified for the last pair: Recovered plaintext matches original.")
else:
    print("\n Decryption Verification Failed for the last pair.")

print("\nPlaintext-Ciphertext Pairs Dictionary:")
print(plaintext_ciphertext_pairs)

KeyError: 2647

In [14]:
import numpy as np

# S-Box for Substitution
s_box = {
    0x0: 0x9, 0x1: 0x4, 0x2: 0xA, 0x3: 0xB,
    0x4: 0xD, 0x5: 0x1, 0x6: 0x8, 0x7: 0x5,
    0x8: 0x6, 0x9: 0x2, 0xA: 0x0, 0xB: 0x3,
    0xC: 0xC, 0xD: 0xE, 0xE: 0xF, 0xF: 0x7
}

# Inverse S-Box
inv_s_box = {v: k for k, v in s_box.items()}

# Round Constants (Rcon)
rcon1 = 0b10000000  # 0x80
rcon2 = 0b00110000  # 0x30

def sub_nibble(byte):
    byte &= 0xFF  # Ensure it's an 8-bit value
    high_nibble = (byte >> 4) & 0x0F
    low_nibble = byte & 0x0F
    return (s_box[high_nibble] << 4) | s_box[low_nibble]


def inv_sub_nibble(byte):
    return (inv_s_box[byte >> 4] << 4) | inv_s_box[byte & 0x0F]

def key_expansion(key):
    w0 = (key >> 8) & 0xFF
    w1 = key & 0xFF

    # Apply S-Box substitution and rotation to w1
    sub_rot_w1 = ((s_box[w1 & 0xF] << 4) | s_box[w1 >> 4])

    # Generate round keys
    w2 = w0 ^ sub_rot_w1 ^ rcon1
    w3 = w2 ^ w1

    sub_rot_w3 = ((s_box[w3 & 0xF] << 4) | s_box[w3 >> 4])

    w4 = w2 ^ sub_rot_w3 ^ rcon2
    w5 = w4 ^ w3

    k1 = (w2 << 8) | w3
    k2 = (w4 << 8) | w5

    return key, k1, k2

def add_round_key(state, key):
    return state ^ key

def shift_rows(state):
    return ((state & 0xF0) | ((state & 0x0C) >> 2) | ((state & 0x03) << 2))

def inv_shift_rows(state):
    return ((state & 0xF0) | ((state & 0x03) << 2) | ((state & 0x0C) >> 2))

def mix_columns(state):
    s0 = (state >> 4) & 0xF
    s1 = state & 0xF

    new_s0 = (s0 ^ multiply_by_4(s1)) & 0xF
    new_s1 = (multiply_by_4(s0) ^ s1) & 0xF

    return (new_s0 << 4) | new_s1

def multiply_by_4(x):
    return ((x << 2) ^ ((x >> 2) & 0x3)) & 0xF

def inv_mix_columns(state):
    s0 = (state >> 4) & 0xF
    s1 = state & 0xF

    new_s0 = (multiply_by_9(s0) ^ multiply_by_2(s1)) & 0xF
    new_s1 = (multiply_by_2(s0) ^ multiply_by_9(s1)) & 0xF

    return (new_s0 << 4) | new_s1

def multiply_by_2(x):
    return ((x << 1) ^ ((x >> 3) & 0x1)) & 0xF

def multiply_by_9(x):
    return (multiply_by_2(multiply_by_2(multiply_by_2(x))) ^ x) & 0xF

def encrypt(plaintext, key):
    k0, k1, k2 = key_expansion(key)


    state = sub_nibble(state)
    print(f"After SubNibble: {state:08b}")
    state = shift_rows(state)
    print(f"After ShiftRows: {state:08b}")
    state = mix_columns(state)
    print(f"After MixColumns: {state:08b}")
    state = add_round_key(state, k1)
    print(f"After AddRoundKey: {state:08b}")
    state = sub_nibble(state)
    print(f"After SubNibble: {state:08b}")
    state = shift_rows(state)
    print(f"After ShiftRows: {state:08b}")
    state = add_round_key(state, k2)
    print(f"After AddRoundKey: {state:08b}")

    return state

def decrypt(ciphertext, key):
    k0, k1, k2 = key_expansion(key)

    state = add_round_key(ciphertext, k2)
    print(f"After AddRoundKey: {state:08b}")
    state = inv_shift_rows(state)
    print(f"After ShiftRows: {state:08b}")
    state = inv_sub_nibble(state)
    print(f"After SubNibble: {state:08b}")
    state = add_round_key(state, k1)
    state = inv_mix_columns(state)
    print(f"After MixColumns: {state:08b}")
    state = inv_shift_rows(state)
    print(f"After ShiftRows: {state:08b}")
    state = inv_sub_nibble(state)
    print(f"After SubNibble: {state:08b}")


    return state

# Test Example
plaintext = 0b11001100  # Example 8-bit plaintext
key = 0b1010010110110101  # Example 16-bit key

ciphertext = encrypt(plaintext, key)
decrypted_text = decrypt(ciphertext, key)

print(f"Plaintext: {bin(plaintext)}")
print(f"Ciphertext: {bin(ciphertext)}")
print(f"Decrypted: {bin(decrypted_text)}")



UnboundLocalError: cannot access local variable 'state' where it is not associated with a value